File: harness/ramp_colour_verify.py
"""Python side of the app-port check: the colour-guided collapse on each picture's RAW 16-bit map (what the 16-bit path
hands the app's collapse) in both modes; writes ref_<pic>_<mode>.f32 for ramp_colour_verify.js to compare against."""
import os, sys, json, numpy as np
from PIL import Image
sys.path.insert(0, '/home/user/moebiusv2/harness')
from ramp_colour import collapse_colour
from ramp_collapse import visible_step
OUT = sys.argv[1]; os.makedirs(OUT, exist_ok=True); R = '/home/user/moebiusv2'
PIC = {'troll': ('defaultImgDepth.png', 'defaultImgColor.png'), 'vermeer': ('harness/batchB/vermeer_da3_16.png', 'harness/batchB/vermeer_color.png'),
       'sunflowers': ('harness/batchB/room_da3_16.png', 'harness/batchB/room_color.png'), 'starwatcher': ('harness/batchB/starwatcher_da3_16.png', 'harness/batchB/starwatcher_color.png')}
res = {}
for p, (dp, cp) in PIC.items():
    d = (np.asarray(Image.open(os.path.join(R, dp))).astype(np.float64) / 65535).astype(np.float32); ph, pw = d.shape
    rgb = np.asarray(Image.open(os.path.join(R, cp)).convert('RGB').resize((pw, ph)), np.float64)
    st = visible_step(pw, ph, 0.02, 0.04, 0.2); d.tofile(os.path.join(OUT, 'in_%s.f32' % p)); np.asarray(Image.open(os.path.join(R, cp)).convert('RGBA').resize((pw, ph)), np.uint8).tofile(os.path.join(OUT, 'rgba_%s.u8' % p))
    for mode, se in (('strong', False), ('safe', True)):
        c, n, stt = collapse_colour(d.astype(np.float64), rgb, 0.02, 0.04, 0.5, 0.2, st, single_edge=se)
        c.astype(np.float32).tofile(os.path.join(OUT, 'ref_%s_%s.f32' % (p, mode))); res['%s_%s' % (p, mode)] = {'changed': n, 'stats': stt}
    res[p] = {'pw': pw, 'ph': ph, 'step': st}
    print(p, {k: v for k, v in res.items() if k.startswith(p + '_')}, flush=True)
json.dump(res, open(os.path.join(OUT, 'ref.json'), 'w'), indent=1)

File: harness/ramp_colour_verify.js
// Verifies the APP's bgRampColourCollapse (its source extracted from ../moebius.js, so what is tested is what ships)
// against harness/ramp_colour.py on the same inputs (ramp_colour_verify.py writes them). node ramp_colour_verify.js <dir>
'use strict';
const fs = require('fs'), path = require('path'); const dir = process.argv[2];
const src = fs.readFileSync(path.join(__dirname, '..', 'moebius.js'), 'utf8');
const a = src.indexOf('function bgRampColourCollapse('), b = src.indexOf('\nfunction bgRimLawFor(', a);
const bgRampColourCollapse = new Function(src.slice(a, b) + '\nreturn bgRampColourCollapse;')();
const ref = JSON.parse(fs.readFileSync(path.join(dir, 'ref.json')));
const f32 = (p) => { const x = fs.readFileSync(p); return new Float32Array(x.buffer, x.byteOffset, x.byteLength / 4); };
let ok = true;
for (const p of ['troll', 'vermeer', 'sunflowers', 'starwatcher']) {
    const { pw, ph, step } = ref[p]; const d = f32(path.join(dir, 'in_' + p + '.f32')); const rgba = fs.readFileSync(path.join(dir, 'rgba_' + p + '.u8'));
    for (const mode of ['strong', 'safe']) {
        const t0 = Date.now(); const r = bgRampColourCollapse(d, rgba, pw, ph, { outer: 0.02, inner: 0.04, pn: 0.5, D: 0.2 }, step, mode === 'safe'); const ms = Date.now() - t0;
        const py = f32(path.join(dir, 'ref_' + p + '_' + mode + '.f32')); let mx = 0, nd = 0; for (let i = 0; i < py.length; i++) { const e = Math.abs(py[i] - r.out[i]); if (e > mx) mx = e; if (e > 0) nd++; }
        const pr = ref[p + '_' + mode]; const same = pr.changed === r.changed && JSON.stringify(Object.keys(pr.stats).sort().map(k => pr.stats[k])) === JSON.stringify(Object.keys(pr.stats).sort().map(k => r.stats[k] || 0));
        console.log(p, mode, 'changed js', r.changed, 'py', pr.changed, '| stats equal', same, '| max |js-py|', mx.toExponential(2), 'in steps', (mx / step).toExponential(2), '| texels differing', nd, '|', ms, 'ms');
        if (!same || mx / step > 1e-3) ok = false;
    }
}
console.log(ok ? 'PORT VERIFIED' : 'PORT MISMATCH'); process.exit(ok ? 0 : 1);
